### PR TITLE
Relion5 input

### LINKED ISSUE #190: RELION-5 input compatibility
Hi! I just saw the 0.7 release which looks awesome!

I see that output is now compatible with RELION-5, which is great! I was wondering if there are any plans to make input compatible with RELION-5 as well? i.e. one could give a `tomograms.star` file and `pytom-match-pick` could figure out tilt angles, defocus per tilt, dose etc from there by itself?

I can make a PR if I ever get there but I won't promise since I am already delayed in delivering the Scipion plugin 😬 
Thought it didn't hurt to ask just in case anyone is already planning something similar 😃

### PR DESCRIPTION
**Pull request by McHaillet**
Closes #190

@Phaips @rdrighetto also adding you here, in case you have ideas about the implementation for this.

I made a PR to tackle the RELION5 metadata input. I tested it on a RELION5 project I made from [the frames of the tutorial](https://dataverse.nl/dataset.xhtml?persistentId=doi:10.34894/OLYEFI). As it is now, it runs and gives expected results. It does make the input quite easy as the job can be started like this and all 3D-CTF data is fetched from the relion metadata:

`pytom_match_template.py -t Templates/80S_14A.mrc -m Templates/mask.mrc -v Tomograms/job006/tomograms/rec_tomo200528_102.mrc  --particle-diameter 300 --relion5-tomograms-star Tomograms/job006/tomograms.star -d TemplateMatching/ -g 0`

If one would want to loop over all the tomograms in the RELION directory you could do something like this in bash:

`for tomo in Tomograms/job006/tomograms/*.mrc; do pytom_match_template.py -t Templates/80S_14A.mrc -m Templates/mask.mrc -v "$tomo" --particle-diameter 300 --relion5-tomograms-star Tomograms/job006/tomograms.star -d TemplateMatching/ -g 0; done`

This PR is still a draft because I need to add unit tests and tackle the defocus handedness parameter from RELION. I am not sure if it actually applies a flip to the images. 

It also made me aware that the code could be improved using some sort of TiltSeries struct internally that associates all the tilt angles, CTF and dose. But, better left for later to not convolute this.

### CODE REVIEW

**Comment by McHaillet** [comment]
I added unittests and fixed reading of the tomogram handedness. It's ready for review :+1:

File: tests/test_io.py [hunk 1]
```diff
@@ -85,3 +88,21 @@ def test_cast_warning(self):
...
+        self.assertIsInstance(meta_data[3], list)
+        self.assertIsInstance(meta_data[3][0], dict)
+        self.assertIsInstance(meta_data[4], int)
+
+        # test name mismatch
+        tomogram = pathlib.Path("tomogram.mrc")
+        with self.assertRaises(
+            ValueError, msg="Unmatching tomograms name should raise an error."
+        ):
+            parse_relion5_star_data(RELION5_TOMOGRAMS_STAR, tomogram)
```
Comment (sroet) [inline tests/test_io.py:108]: please add any/all of the following tests:
```suggestion
        tomogram = pathlib.Path("tomogram.mrc")
        with self.assertRaises(
            ValueError, msg="Unmatching tomograms name should raise an error."
        ):
            parse_relion5_star_data(RELION5_TOMOGRAMS_STAR, tomogram)

        tomogram = pathlib.Path("rec_tomogram200528_1077.mrc")
        with self.assertRaises(
            ValueError, msg="Partially matching tomogram name should raise an error."
        ):
            parse_relion5_star_data(RELION5_TOMOGRAMS_STAR, tomogram)


        tomogram = pathlib.Path("rec_tomogram200528_10.mrc")
        with self.assertRaises(
            ValueError, msg="Partially matching tomogram name should raise an error."
        ):
            parse_relion5_star_data(RELION5_TOMOGRAMS_STAR, tomogram)
```

File: src/pytom_tm/extract.py [hunk 2]
```diff
@@ -276,6 +276,11 @@ def extract_particles(
     pixel_size = job.voxel_size
     tomogram_id = job.tomo_id
 
+    # remove relion5 reconstructed tomogram name as it messes with linking the tilt
+    # series id when extracting subtomos
+    if relion5_compat and tomogram_id.startswith("rec_"):
+        tomogram_id = tomogram_id[4:]
```
Comment (sroet) [inline src/pytom_tm/extract.py:282]: this added line is not hit by the tests

**Review by sroet — CHANGES_REQUESTED** [review]
1 missing test, couple other issues/corner cases. LGTM otherwise

File: pyproject.toml [hunk 3]
```diff
@@ -4,7 +4,7 @@ build-backend = "setuptools.build_meta"
 
 [project]
 name = "pytom-match-pick"
-version = "0.7.3"
+version = "0.8.0"
```
[the remark this answers is not in the document]
Comment (McHaillet) [reply pyproject.toml:7]: True!

File: src/pytom_tm/io.py [hunk 4]
```diff
@@ -481,3 +482,103 @@ def read_defocus_file(file_name: pathlib.Path) -> list[float, ...]:
...
+        tuple[float, list[float, ...], list[float, ...], list[dict, ...], int]
+    """
+    tomogram_id = tomogram_path.stem
+    tomograms_star_data = starfile.read(tomograms_star_path)
+
+    # match the tomo_id and check if viable
+    matches = [
+        (i, x)
+        for i, x in enumerate(tomograms_star_data["rlnTomoName"])
+        if x in tomogram_id
```
[the remark this answers is not in the document]
Comment (McHaillet) [reply src/pytom_tm/io.py:518]: good point, the second suggestion is perfect-

File: src/pytom_tm/io.py [hunk 5]
```diff
@@ -481,3 +482,103 @@ def read_defocus_file(file_name: pathlib.Path) -> list[float, ...]:
...
+    ]
+    if len(matches) == 1:
+        tomogram_meta_data = tomograms_star_data.loc[matches[0][0]]
+    else:
+        raise ValueError(
+            "Multiple or zero matches of tomogram id in RELION5 STAR file. Aborting..."
+        )
```
[the remark this answers is not in the document]
Comment (McHaillet) [reply src/pytom_tm/io.py:525]: Nice!

File: src/pytom_tm/io.py [hunk 6]
```diff
@@ -481,3 +482,103 @@ def read_defocus_file(file_name: pathlib.Path) -> list[float, ...]:
...
+            "spherical_aberration": tomogram_meta_data["rlnSphericalAberration"] * 1e-3,
+            "flip_phase": phase_flip_correction,
+            "phase_shift_deg": phase_shift,  # Unclear to me where the phase
+            # shifts are annotated in RELION5 tilt series file
+        }
+        for defocus in (
+            tilt_series_star_data.rlnDefocusV + tilt_series_star_data.rlnDefocusU
+        )
+        / 2
+        * 1e-4
```
[the remark this answers is not in the document]
Comment (McHaillet) [reply src/pytom_tm/io.py:565]: Yea it is supposed be that (I have the units in meters). The value in the relion file are in Angstrom, hence the factor 1e-10. I will change it to do it in one go, because this reads weird.

File: src/pytom_tm/extract.py (hunk 2 above)
Comment (McHaillet) [reply src/pytom_tm/extract.py:282]: Made sure its hit now!

**Review by sroet — APPROVED** [review]
Thanks for the fixes, LGTM
